commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+          * of the removed entry.
+          *
+          * @param removed removed entry view
+          */
+         default void onRemove(ReadEntryView<K, V> removed) {}
+      }
+   }
+
+   /**
+    * Write-only listeners enables user to register listeners for events
+    * happening in the write-only functional map.
+    *
+    * DESIGN RATIONALES:
+    * <ul>
+    *    <li>Write-only listeners cannot distinguish between cache entry
+    *    created and cache entry modify/update events. In either case, all
+    *    they know is that a new entry has been added, but they can't know if
+    *    entry was present before or not, nor what the previous value.
+    *    </li>
+    *    <li>Write-only listeners can distinguish between entry removals
+    *    vs cache entry create/modify-update events because they can query
+    *    what the new entry's value via {@link ReadEntryView#find()}.
+    *    </li>
+    * </ul>
+    */
+   public interface WriteListeners<K, V> {
+      /**
+       * Add a write event listener by passing in a {@link Consumer} to be
+       * called each time a cache entry is created, modified/updated or
+       * removed.
+       *
+       * For created or modified/updated events, the
+       * {@link ReadEntryView} passed in will represent the newly stored
+       * entry, hence implementations will not be available to differentiate
+       * between created events vs modified/updated events.
+       *
+       * For removed events, {@link ReadEntryView} passed in will represent
+       * an empty entry view, hence {@link ReadEntryView#find()} will return
+       * an {@link java.util.Optional} instance that's empty, and
+       * {@link ReadEntryView#get()} will throw
+       * {@link java.util.NoSuchElementException}.
+       *
+       * @param f operation to be called each time a cache entry is written
+       * @return an {@link AutoCloseable} instance that can be used to
+       *         unregister the listener
+       */
+      AutoCloseable onWrite(Consumer<ReadEntryView<K, V>> f);
+
+      /**
+       * Add a write-only listener, and return an {@link AutoCloseable}
+       * instance that can be used to remove the listener registration.
+       *
+       * @param l the write-only functional map event listener
+       * @return an {@link AutoCloseable} instance that can be used to
+       *         unregister the listener
+       */
+      AutoCloseable add(WriteListener<K, V> l);
+
+      /**
+       * Write-only listener.
+       */
+      interface WriteListener<K, V> {
+         /**
+          * Entry write event callback that receives a {@link ReadEntryView}
+          * of the written entry.
+          *
+          * For created or modified/updated events, the
+          * {@link ReadEntryView} passed in will represent the newly stored
+          * entry, hence implementations will not be available to differentiate
+          * between created events vs modified/updated events.
```

Comment:
Might mention that the listener may be called even if the entry didn't actually change, e.g. if there's a cache store attached we might not read the previous value.